File: examples/running/transport/handler.go
// tb-cell: handlers go-example 🟡 -- v3 transport/ shape; adapters/handlers layout pending the Go mirror
package transport

import (
	"encoding/json"
	"net/http"

	"github.com/verocorp/tesser-build/examples/running/linkcampaign"
)

type Handler struct {
	client linkcampaign.Client
	mux    *http.ServeMux
}

func NewHandler(client linkcampaign.Client) *Handler {
	h := &Handler{client: client, mux: http.NewServeMux()}
	h.mux.HandleFunc("POST /campaigns", h.createCampaign)
	h.mux.HandleFunc("GET /campaigns/{id}", h.getCampaign)
	h.mux.HandleFunc("POST /campaigns/{id}/links", h.addShortLink)
	h.mux.HandleFunc("POST /campaigns/{id}/links/{slug}/deactivate", h.deactivateShortLink)
	return h
}

func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	h.mux.ServeHTTP(w, r)
}

func (h *Handler) createCampaign(w http.ResponseWriter, r *http.Request) {
	var req linkcampaign.CreateCampaignRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, err)
		return
	}
	resp, err := h.client.CreateCampaign(r.Context(), req)
	if err != nil {
		writeError(w, http.StatusUnprocessableEntity, err)
		return
	}
	writeJSON(w, http.StatusCreated, resp)
}

func (h *Handler) getCampaign(w http.ResponseWriter, r *http.Request) {
	req := linkcampaign.GetCampaignRequest{CampaignID: r.PathValue("id")}
	resp, err := h.client.GetCampaign(r.Context(), req)
	if err != nil {
		writeError(w, http.StatusNotFound, err)
		return
	}
	writeJSON(w, http.StatusOK, resp)
}

func (h *Handler) addShortLink(w http.ResponseWriter, r *http.Request) {
	var body linkcampaign.ShortLinkInput
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, err)
		return
	}
	req := linkcampaign.AddShortLinkRequest{
		CampaignID: r.PathValue("id"),
		Slug:       body.Slug,
		TargetURL:  body.TargetURL,
	}
	resp, err := h.client.AddShortLink(r.Context(), req)
	if err != nil {
		writeError(w, http.StatusUnprocessableEntity, err)
		return
	}
	writeJSON(w, http.StatusOK, resp)
}

func (h *Handler) deactivateShortLink(w http.ResponseWriter, r *http.Request) {
	req := linkcampaign.DeactivateShortLinkRequest{
		CampaignID: r.PathValue("id"),
		Slug:       r.PathValue("slug"),
	}
	resp, err := h.client.DeactivateShortLink(r.Context(), req)
	if err != nil {
		writeError(w, http.StatusUnprocessableEntity, err)
		return
	}
	writeJSON(w, http.StatusOK, resp)
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func writeError(w http.ResponseWriter, status int, err error) {
	writeJSON(w, status, map[string]string{"error": err.Error()})
}

File: examples/python/transport/handler.py
import json
import re
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler
from typing import Any

from linkcampaign import (
    AddShortLinkRequest,
    Client,
    CreateCampaignRequest,
    DeactivateShortLinkRequest,
    GetCampaignRequest,
    ShortLinkInput,
)

_CAMPAIGNS = re.compile(r"^/campaigns$")
_CAMPAIGN = re.compile(r"^/campaigns/(?P<id>[^/]+)$")
_LINKS = re.compile(r"^/campaigns/(?P<id>[^/]+)/links$")
_DEACTIVATE = re.compile(r"^/campaigns/(?P<id>[^/]+)/links/(?P<slug>[^/]+)/deactivate$")


def make_handler(client: Client) -> type[BaseHTTPRequestHandler]:

    class Handler(BaseHTTPRequestHandler):
        def do_POST(self) -> None:  # noqa: N802 (stdlib dispatch name)
            if _CAMPAIGNS.match(self.path):
                self._create_campaign()
            elif (m := _LINKS.match(self.path)) is not None:
                self._add_short_link(m.group("id"))
            elif (m := _DEACTIVATE.match(self.path)) is not None:
                self._deactivate_short_link(m.group("id"), m.group("slug"))
            else:
                self._write_error(HTTPStatus.NOT_FOUND, "no such route")

        def do_GET(self) -> None:  # noqa: N802 (stdlib dispatch name)
            if (m := _CAMPAIGN.match(self.path)) is not None:
                self._get_campaign(m.group("id"))
            else:
                self._write_error(HTTPStatus.NOT_FOUND, "no such route")


        def _create_campaign(self) -> None:
            body = self._read_json()
            if body is None:
                return
            req = CreateCampaignRequest(
                name=str(body.get("name", "")),
                links=tuple(
                    ShortLinkInput(
                        slug=str(l.get("slug", "")),
                        target_url=str(l.get("target_url", "")),
                    )
                    for l in _as_list(body.get("links"))
                ),
            )
            try:
                resp = client.create_campaign(req)
            except ValueError as e:
                return self._write_error(HTTPStatus.UNPROCESSABLE_ENTITY, str(e))
            self._write_json(HTTPStatus.CREATED, resp)

        def _get_campaign(self, campaign_id: str) -> None:
            try:
                resp = client.get_campaign(GetCampaignRequest(campaign_id=campaign_id))
            except (ValueError, LookupError) as e:
                return self._write_error(HTTPStatus.NOT_FOUND, str(e))
            self._write_json(HTTPStatus.OK, resp)

        def _add_short_link(self, campaign_id: str) -> None:
            body = self._read_json()
            if body is None:
                return
            req = AddShortLinkRequest(
                campaign_id=campaign_id,
                slug=str(body.get("slug", "")),
                target_url=str(body.get("target_url", "")),
            )
            try:
                resp = client.add_short_link(req)
            except (ValueError, LookupError) as e:
                return self._write_error(HTTPStatus.UNPROCESSABLE_ENTITY, str(e))
            self._write_json(HTTPStatus.OK, resp)

        def _deactivate_short_link(self, campaign_id: str, slug: str) -> None:
            req = DeactivateShortLinkRequest(campaign_id=campaign_id, slug=slug)
            try:
                resp = client.deactivate_short_link(req)
            except (ValueError, LookupError) as e:
                return self._write_error(HTTPStatus.UNPROCESSABLE_ENTITY, str(e))
            self._write_json(HTTPStatus.OK, resp)


        def _read_json(self) -> dict[str, Any] | None:
            length = int(self.headers.get("Content-Length", 0))
            raw = self.rfile.read(length)
            try:
                payload = json.loads(raw)
            except json.JSONDecodeError as e:
                self._write_error(HTTPStatus.BAD_REQUEST, f"invalid JSON: {e}")
                return None
            if not isinstance(payload, dict):
                self._write_error(HTTPStatus.BAD_REQUEST, "expected a JSON object")
                return None
            return payload

        def _write_json(self, status: HTTPStatus, body: object) -> None:
            data = json.dumps(_as_jsonable(body)).encode()
            self.send_response(status)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(data)))
            self.end_headers()
            self.wfile.write(data)

        def _write_error(self, status: HTTPStatus, message: str) -> None:
            self._write_json(status, {"error": message})

        def log_message(self, format: str, *args: Any) -> None:
            pass

    return Handler


def _as_list(value: object) -> list[Any]:
    return value if isinstance(value, list) else []


def _as_jsonable(body: object) -> object:
    from dataclasses import asdict, is_dataclass

    if is_dataclass(body) and not isinstance(body, type):
        return asdict(body)
    return body
